utils/frame/benchmarking-cli/src/storage/write.rs

```diff
@@ -0,0 +1,129 @@
+// limitations under the License.
+
+use sc_cli::Result;
+use sc_client_api::UsageProvider;
+use sc_client_db::{columns, DB_HASH_LEN, DatabaseSource, DbHash, DbState};
+use sc_service::Configuration;
+use sp_api::StateBackend;
+use sp_blockchain::HeaderBackend;
+use sp_core::H256;
+use sp_trie::PrefixedMemoryDB;
+use sp_database::Transaction;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, HashFor, Header as HeaderT},
+};
+
+use log::info;
+use rand::prelude::*;
+use std::{fmt::Debug, sync::Arc, time::Instant};
+
+use super::{cmd::StorageCmd, record::BenchRecord};
+
+impl StorageCmd {
+	/// Benchmarks the time it takes to write a single Storage item.
+	/// Uses the latest state that is available for the given client.
+	pub(crate) fn bench_write<Block, H, C>(
+		&self,
+		cfg: &Configuration,
+		client: Arc<C>,
+		db: Arc<dyn sp_database::Database<DbHash>>,
+		storage: Arc<dyn sp_state_machine::Storage<HashFor<Block>>>,
+	) -> Result<BenchRecord>
+	where
+		Block: BlockT<Header = H, Hash = H256> + Debug,
+		H: HeaderT<Hash = H256>,
+		C: UsageProvider<Block> + HeaderBackend<Block>,
+	{
+		// Store the time that it took to write each value.
+		let mut record = BenchRecord::default();
+
+		let is_parity = matches!(cfg.database, DatabaseSource::ParityDb { path: _ });
+		let block = BlockId::Number(client.usage_info().chain.best_number);
+		let header = client.header(block)?.ok_or("Header not found")?;
+		let original_root = *header.state_root();
+		let trie = DbState::<Block>::new(storage.clone(), original_root);
+
+		info!("Preparing keys from block {}", block);
+		// Load all KV pairs and randomly shuffle them.
+		let mut kvs = trie.pairs();
+		let mut rng = self.setup_rng();
+		kvs.shuffle(&mut rng);
+
+		info!("Writing {} keys", kvs.len());
+		// Write each value in one commit.
+		for (k, original_v) in kvs.iter() {
+			// Create a random value to overwrite with.
+			// NOTE: We use a possibly higher entropy than the original value,
+			// could be improved but acts as an over-estimation which is fine for now.
+			let mut new_v = vec![0; original_v.len()];
+			rng.fill_bytes(&mut new_v[..]);
+
+			// Interesting part here:
+			let start = Instant::now();
+			// Create a TX that will modify the Trie in the DB and
+			// calculate the root hash of the Trie after the modification.
+			let replace = vec![(k.as_ref(), Some(new_v.as_ref()))];
+			let (root, tx) = trie.storage_root(replace.iter().cloned(), self.state_version());
+			let tx = convert_tx::<Block>(tx, is_parity);
+			db.commit(tx).map_err(|e| format!("Writing to the Database: {}", e))?;
+
+			record.append(new_v.len(), start.elapsed())?;
+
+			// Now undo the change:
+			// Create a Trie with the modified root hash and replace the value with its original.
+			let trie = DbState::<Block>::new(storage.clone(), root);
+			let replace = vec![(k.as_ref(), Some(original_v.as_ref()))];
+			let (root, tx) = trie.storage_root(replace.iter().cloned(), self.state_version());
```

Comment:
There seems to be a small chance that this corrupts a DB here.  
Personally I would leave it for now and add a `--i-used-a-backup` flag to the CLI.  

To quote @cheme:  
>rocksdb : risk of collision for small value when removing value created in new state
>paritydb : fine (except if version mismatch)

Reply:
@cheme what?

Reply:
The method to restore the state can delete a node that was in an earlier state.
Eg at block 0 you got key1 value 0, then at block 1 you got key1 value 1 then we run the bench and it change key 1 to value 0 (one chance on 256) : this will write an encoded node that already exists in the db so noops, but when reverting it state after bench, it will remove the node and state at block 0 would be corrupted.
On paritydb reference counted are used so we won't have the issue.